test/python/circuit/library/test_linear_function.py

```diff
@@ -0,0 +1,224 @@
+        """Test correctness of first synthesizing a linear circuit from a linear function,
+        and then converting this linear circuit to a linear function."""
+        rng = np.random.default_rng(5678)
+
+        for _ in range(10):
+            # create a random invertible binary matrix
+            binary_matrix = random_invertible_binary_matrix(num_qubits, seed=rng)
+
+            # create a linear function with this matrix
+            linear_function = LinearFunction(binary_matrix, validate_input=True)
+            self.assertTrue(np.all(linear_function.linear == binary_matrix))
+
+            # synthesize linear function
+            synthesized_circuit = linear_function.definition
+            self.assertTrue(isinstance(synthesized_circuit, QuantumCircuit))
+
+            # check that the synthesized linear function only contains CX and SWAP gates
+            for inst, _, _ in synthesized_circuit.data:
+                self.assertTrue(isinstance(inst, (CXGate, SwapGate)))
+
+            # construct a linear function out of this linear circuit
+            synthesized_linear_function = LinearFunction(synthesized_circuit, validate_input=True)
+
+            # check equivalence of the two linear matrices
+            self.assertTrue(np.all(synthesized_linear_function.linear == binary_matrix))
+
+    def test_patel_markov_hayes(self):
+        """Checks the explicit example from Patel-Markov-Hayes's paper."""
+
+        # This code is adapted from test_synthesis.py
+        binary_matrix = [
+            [1, 1, 0, 0, 0, 0],
+            [1, 0, 0, 1, 1, 0],
+            [0, 1, 0, 0, 1, 0],
+            [1, 1, 1, 1, 1, 1],
+            [1, 1, 0, 1, 1, 1],
+            [0, 0, 1, 1, 1, 0],
+        ]
+
+        # Construct linear function from matrix: here, we copy a matrix
+        linear_function_from_matrix = LinearFunction(binary_matrix, validate_input=True)
+
+        # Create the circuit displayed above:
+        linear_circuit = QuantumCircuit(6)
+        linear_circuit.cx(4, 3)
+        linear_circuit.cx(5, 2)
+        linear_circuit.cx(1, 0)
+        linear_circuit.cx(3, 1)
+        linear_circuit.cx(4, 2)
+        linear_circuit.cx(4, 3)
+        linear_circuit.cx(5, 4)
+        linear_circuit.cx(2, 3)
+        linear_circuit.cx(3, 2)
+        linear_circuit.cx(3, 5)
+        linear_circuit.cx(2, 4)
+        linear_circuit.cx(1, 2)
+        linear_circuit.cx(0, 1)
+        linear_circuit.cx(0, 4)
+        linear_circuit.cx(0, 3)
+
+        # Construct linear function from matrix: we build a matrix
+        linear_function_from_circuit = LinearFunction(linear_circuit, validate_input=True)
+
+        # Compare the matrices
+        self.assertTrue(
+            np.all(linear_function_from_circuit.linear == linear_function_from_matrix.linear)
+        )
+
+        self.assertTrue(
+            Operator(linear_function_from_matrix.definition) == Operator(linear_circuit)
+        )
```

Comment:
It would be good to check some outputs of this function against explicit circuits in circuit form, where we can check that the circuit definition is exactly what we expect.  So far, these tests check that the synthesis is valid, but not that it's efficient.

Reply:
thanks! hopefully answered this in the comment below